docs/design/batch-request.md

```diff
@@ -0,0 +1,260 @@
+            "name": "gasLimit too high",
+            "message": "[Request ID: 9e2c40fb-3406-4c48-b7ac-7a40e9bdb880] Transaction gas limit '0xe1f21c67' exceeds block gas limit '15000000'"
+        },
+        "jsonrpc": "2.0",
+        "id": 4
+    }
+]
+```
+
+### Bad Request on the Batch Request as a whole
+This type of error happens when the whole batch request fails and could happen for the following reasons:
+1. The request is not a valid JSON object or Array of objects
+```json
+{
+    "error": {
+        "message": "Parse error",
+        "code": -32700
+    },
+    "jsonrpc": "2.0",
+    "id": null
+}
+```
+2. The feature is disabled
+```json
+{
+    "error": {
+        "code": -32202,
+        "name": "Batch requests disabled",
+        "message": "Batch requests are disabled"
+    },
+    "jsonrpc": "2.0",
+    "id": null
+}
+```
+3. The request Array size is bigger than the configured limit
+
+```json
+{
+    "jsonrpc": "2.0",
+    "error": {
+        "code": -32005,
+        "message": "batch item count exceeded"
+    }
+}
+```
+
+## Metrics Capture Enhancement
+Currently metrics are captured using a KoaApp middleware that captures the metric `rpc_relay_method_response` per each HTTP request capturing the amount of time it took to process the request in milliseconds (ms).
+
+The metric currently uses the following labels:
+
+1. `method` (method name such as eth_call, eth_getBlockByNumber or batch_request)
+2. `status` (HTTP status code)
+
+This approach works for both single and batch requests, however it does not capture the metrics for each individual request within the batch request.
+
+Given the RPC team's heavy reliance on this metric to measure the performance and reliability of the RPC service, it is crucial to maintain its integrity.
+
+### Proposed Solution
+The suggested enhancement is to retain the existing metric unchanged while introducing an additional metric to track individual requests within a batch.
+
+The new metric will be called `rpc_relay_batch_request` and will capture the following labels:
+1. `method` (method name such as eth_call, eth_getBlockByNumber)
+2. `status` (instead of HTTP status code, it will be the status of the individual request, success or error, in case of sucess it will have the value `200` to keep it similar to the current metric, but in case of error it will have the RPC error code like `-32005` or `-32006`)
+
+The processing time in milliseconds (ms) will continue to be observed for this metric as well.
+
+This proposed approach offers several benefits. It preserves the integrity of our current metric system while enabling the capture of detailed performance metrics for each request within a batch. This data can then be showcased on a new dashboard designed to monitor method performance independently of request type—singular or batch. Concurrently, our existing dashboards will continue to provide comprehensive metrics for all single and batch requests at the whole HTTP request level.
```

Comment:
Q: does this mean the dashboard would need an aggregated view of both the metrics?

Reply:
Yes, but only if what we want to measure is the time it takes a `method` to reply, if we are measuring the time it takes an `http request` to reply, the same dashboards will work.

So My initial thought is to create a new dashboard and be more descriptive on the intended use for each one, in one we are looking for http latency (potencial timeouts) and in the other we are measuring the time it takes a method to complete. (they will overlap in the case of single request)

If you have other ideas am open to discussing them. 👍

Reply:
✅  I was able to find another approach that might also work better and simpler 👍